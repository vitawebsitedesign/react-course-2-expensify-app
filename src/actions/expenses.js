import {uuid} from 'uuidv4';

export const addExpense = ({
    description = '',
    note = '',
    amt = 0,
    createdAt = new Date()
}) => ({
    type: 'ADD_EXPENSE',
    expense: {
        id: uuid(),
        description,
        note,
        amt,
        createdAt
    }
});

export const removeExpense = (id) => ({
    type: 'REMOVE_EXPENSE',
    id
});

export const editExpense = (id, expense) => ({
    type: 'EDIT_EXPENSE',
    id,
    expense
});
