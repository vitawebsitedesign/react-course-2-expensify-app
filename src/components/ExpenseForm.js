import React from 'react';
import moment from 'moment';
import { SingleDatePicker } from 'react-dates';

export default class ExpenseForm extends React.Component {
    constructor(props) {
        super(props);
        this.state = {
            desc: this.props.expense ? this.props.expense.description : '',
            note: this.props.expense ? this.props.expense.note : '',
            amt: this.props.expense ? (this.props.expense.amt / 100).toString() : '',
            createdAt: (this.props.expense && moment.unix) ? moment.unix(this.props.expense.createdAt / 1000) : moment(),
            calendarFocused: false,
            err: 'Please provide a description and amount'
        };
    }

    onChangeDesc = e => {
        const desc = e.target.value;
        this.setState(() => ({
            desc
        }));
    };

    onChangeNote = e => {
        const note = e.target.value;
        this.setState(() => ({
            note
        }));
    };

    onChangeAmt = e => {
        const amt = e.target.value;
        const amtRegex = /^\d*(\.\d{0,2})?$/;
        const amtValid = amt.match(amtRegex);
        if (amtValid) {
            this.setState(() => ({
                amt
            }));
        }
    };

    onDateChange = createdAt => {
        if (createdAt) {
            this.setState(() => ({
                createdAt
            }));
        }
    };

    onFocusChange = ({ focused }) => {
        this.setState(() => ({
            calendarFocused: focused
        }));
    };

    onSubmit = e => {
        e.preventDefault();
        const {desc, amt, note, createdAt} = this.state;

        const err = this.onSubmitValidate(desc, amt);
        if (!err) {
            this.props.onSubmit({
                description: desc,
                amt: parseFloat(amt),
                note: note,
                createdAt: createdAt.valueOf()
            });
        }

        this.setState({
            err
        });
    };

    onSubmitValidate = (desc, amt) => {
        if (!desc) {
            return 'Error: Please provide a description';
        } else if (amt === null || amt === undefined || amt === '') {
            return 'Error: Please provide an amount';
        }
        return null;
    };

    render() {
        return (
            <div>
                {this.state.err && <p data-oa-name="err">{this.state.err}</p>}
                <form onSubmit={this.onSubmit}>
                    <label htmlFor="desc">
                        Description:
                        <input type="text" name="desc" value={this.state.desc} onChange={this.onChangeDesc} autoFocus />
                    </label>
                    <label htmlFor="note">
                        Note (optional):
                        <textarea name="note" value={this.state.note} onChange={this.onChangeNote} />
                    </label>
                    <label htmlFor="amt">
                        Amount:
                        <input type="text" name="amt" value={this.state.amt} onChange={this.onChangeAmt} />
                    </label>
                    <SingleDatePicker
                        id="expense-date-picker"
                        date={this.state.createdAt}
                        onDateChange={this.onDateChange}
                        focused={this.state.calendarFocused}
                        onFocusChange={this.onFocusChange}
                        isOutsideRange={() => false}
                        displayFormat="MMM Do, YYYY"
                    />
                    <button>Create expense</button>
                </form>
            </div>
        );
    }
}
