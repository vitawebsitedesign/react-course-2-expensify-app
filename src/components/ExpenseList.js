import React from 'react';
import {connect} from 'react-redux';
import ExpenseListItem from './ExpenseListItem';
import getVisibleExpenses from '../selectors/expenses';

export const ExpenseList = props => {
    const expenses = props.expenses.map(e =>
        <ExpenseListItem
            key={e.id}
            {...e}
        />
    );

    return (
        <ul>
            {
                expenses.length === 0 ? (
                    <p>No expenses</p>
                ) : (
                    <div>{expenses}</div>
                )
            }
        </ul>
    );
};

const mapStateToProps = state => ({
    expenses: getVisibleExpenses(state.expenses, state.filters)
});

export default connect(mapStateToProps)(ExpenseList);
