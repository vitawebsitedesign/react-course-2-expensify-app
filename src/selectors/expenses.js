import moment from 'moment';

const getVisibleExpenses = (expenses, filters) => {
    const {text, sortBy, startDate, endDate} = filters;
    return expenses.filter(e => {
        const createdAtMoment = moment(e.createdAt);
        const startDateMatch = startDate ? createdAtMoment.isSameOrAfter(startDate, 'day') : true;
        const endDateMatch = endDate ? createdAtMoment.isSameOrBefore(endDate, 'day') : true;
        const textMatch = e.description.toLowerCase().includes(text.toLowerCase());
        return startDateMatch && endDateMatch && textMatch;
    }).sort((a, b) => {
        if (sortBy === 'date') {
            return a.createdAt - b.createdAt;
        } else if (sortBy === 'amt') {
            return b.amt - a.amt;
        }
        return 0;
    });
};

export default getVisibleExpenses;
