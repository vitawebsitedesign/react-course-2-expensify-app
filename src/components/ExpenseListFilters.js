import React from 'react';
import {connect} from 'react-redux';
import {setTextFilter, sortByAmt, sortByDate, setStartDate, setEndDate} from '../actions/filters';
import {DateRangePicker} from 'react-dates';

export class ExpenseListFilters extends React.Component {
    state = {
        calendarFocused: null,
    };

    onDatesChange = ({startDate, endDate}) => {
        this.props.setFilterStartDate(startDate);
        if (endDate) {
            this.props.setFilterEndDate(endDate);
        }
    };

    onFocusChange = calendarFocused => {
        this.setState(() => ({
            calendarFocused
        }))
    };

    onTextChange = e => {
        this.props.setFilterText(e.target.value);
    };

    onSortByChange = e => {
        let sortByAction = null;
        switch (e.target.value) {
            case 'date':
                this.props.setFilterSortByDate();
                break;
            case 'amt':
                this.props.setFilterSortByAmt();
                break;
            default:
                throw new Error('unhandled sort');
        }
    };

    render() {
        return (
            <div>
                <input type="text" onChange={this.onTextChange} value={this.props.filters.text}></input>
                <select value={this.props.filters.sortBy} onChange={this.onSortByChange}>
                    <option value="date">Date</option>
                    <option value="amt">Amount</option>
                </select>
                <div class="wrapper-date-range-picker">
                    <DateRangePicker
                        startDate={this.props.filters.startDate}
                        startDateId="date-range-picker-start-date"
                        endDate={this.props.filters.endDate}
                        endDateId="date-range-picker-end-date"
                        onDatesChange={this.onDatesChange}
                        focusedInput={this.state.calendarFocused}
                        onFocusChange={this.onFocusChange}
                        isOutsideRange={() => false}
                        showClearDates={true}
                    />
                </div>
            </div>
        );
    }
}

const mapStateToProps = state => ({
    filters: state.filters
});

const mapDispatchToProps = dispatch => ({
    setFilterStartDate: startDate => dispatch(setStartDate(startDate)),
    setFilterEndDate: endDate => dispatch(setEndDate(endDate)),
    setFilterText: text => dispatch(setTextFilter(text)),
    setFilterSortByDate: () => dispatch(sortByDate()),
    setFilterSortByAmt: () => dispatch(sortByAmt())
});

export default connect(mapStateToProps, mapDispatchToProps)(ExpenseListFilters);
