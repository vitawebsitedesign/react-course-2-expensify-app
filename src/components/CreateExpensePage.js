import React from 'react';
import ExpenseForm from './ExpenseForm';
import {connect} from 'react-redux';
import {addExpense} from '../actions/expenses';

export class CreateExpensePage extends React.Component {
    onSubmit = e => {
        this.props.addExpense(e);
        this.props.history.push('/');
    };

    render() {
        return (
            <div>
                <p>Use the form below to create a new expense.</p>
                <ExpenseForm onSubmit={this.onSubmit} />
            </div>
        );
    }
}

const mapDispatchToProps = dispatch => ({
    addExpense: e => dispatch(addExpense(e))
});

export default connect(undefined, mapDispatchToProps)(CreateExpensePage);
