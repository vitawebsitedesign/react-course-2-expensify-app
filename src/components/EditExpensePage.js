import React from 'react';
import {connect} from 'react-redux';
import ExpenseForm from './ExpenseForm';
import {editExpense, removeExpense} from '../actions/expenses';

export class EditExpensePage extends React.Component {
    render() {
        return (
            <div>
                <ExpenseForm onSubmit={e => {
                        this.props.editExpense(e);
                        this.props.history.push('/');
                    }}
                    expense={this.props.expense}
                />
                <button onClick={() => {
                    this.props.removeExpense();
                    this.props.history.push('/');
                }} data-oa-name="btn-remove">remove</button>
            </div>
        );
    }
}

const mapStateToProps = (state, props) => ({
    expense: state.expenses.find(e => e.id === props.match.params.id)
});

const mapDispatchToProps = (dispatch, props) => ({
    editExpense: (e) => {
        debugger;
        const expense = {...e};
        expense.amt *= 100;
        dispatch(editExpense(props.match.params.id, expense));
    },
    removeExpense: () => {debugger;dispatch(removeExpense(props.match.params.id));}
});

export default connect(mapStateToProps, mapDispatchToProps)(EditExpensePage);
