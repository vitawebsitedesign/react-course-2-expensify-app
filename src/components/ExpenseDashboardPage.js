import React from 'react';
import ExpenseList from './ExpenseList';
import ExpenseListFilters from './ExpenseListFilters';

const ExpenseDashboardPage = () => (
    <div>
        <p>
            yo, whatup biatch
        </p>
        <ExpenseListFilters />
        <ExpenseList />
    </div>
);

export default ExpenseDashboardPage
