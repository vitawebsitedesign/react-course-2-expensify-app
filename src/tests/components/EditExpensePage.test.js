import React from 'react';
import {shallow} from 'enzyme';
import {EditExpensePage} from '../../components/EditExpensePage';
import {gum} from '../fixtures/expenses';

let editExpenseSpy, historySpy, removeExpenseSpy, wrapper;

beforeEach(() => {
    editExpenseSpy = jest.fn();
    historySpy = {
        push: jest.fn()
    };
    removeExpenseSpy = jest.fn();
    wrapper = shallow(<EditExpensePage expense={gum} editExpense={editExpenseSpy} removeExpense={removeExpenseSpy} history={historySpy} />);
});

test('should render', () => {
    const wrapper = shallow(<EditExpensePage />);
    expect(wrapper).toMatchSnapshot();
});

test('should edit expense', () => {
    const editedExpense = {
        ...gum,
        amt: 405
    };
    const evt = {
        e: editedExpense
    };

    wrapper.find('ExpenseForm').simulate('submit', evt);
    expect(editExpenseSpy).toHaveBeenLastCalledWith(evt);
    expect(historySpy.push).toHaveBeenLastCalledWith('/');
});

test('should remove expense', () => {
    wrapper.find('[data-oa-name="btn-remove"]').simulate('click');
    expect(removeExpenseSpy).toHaveBeenCalledWith();
    expect(historySpy.push).toHaveBeenLastCalledWith('/');
});
