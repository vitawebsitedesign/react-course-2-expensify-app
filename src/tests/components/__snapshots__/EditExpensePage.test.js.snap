// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render 1`] = `
<div>
  <ExpenseForm
    onSubmit={[Function]}
  />
  <button
    data-oa-name="btn-remove"
    onClick={[Function]}
  >
    remove
  </button>
</div>
`;
